### ISSUE TITLE
Data limit for rsc.tl.harmony_integrate() lower than expected

### ISSUE THREAD

**Issue by johnhickey22** [scverse/rapids_singlecell#30]
I am using the rapids platform in a Google Colab Pro Plus with a dataset with about 5.6 million cells, but only 40 markers (so I figured I would be fine on a A100 GPU with 90 GB RAM that I am using). I also had tried running on a T4 but it said that there was not enough memory for the harmony integration (it flagged such an error when I tried to run). I ran a subset (5% of the cells) and it was able to run (and very quickly); however when I try to run the whole dataset I get the following error: 

# CUTLASS error encountered at: file=/__w/cuml/cuml/python/_skbuild/linux-x86_64-3.10/cmake-build/_deps/raft-src/cpp/include/raft/distance/detail/pairwise_distance_cutlass_base.cuh line=159: call='cutlassDist_op.can_implement(arguments)', Reason=Error Invalid Problem
# Obtained 64 stack frames

Any way to get around the memory issue? Or set up my notebook differently? I am able to get other rsc functions running well (e.g., rsc.tl.leiden). Thanks!

**Comment by Intron7** [comment]
Dear John,

Have you tried using rmm? 
```
import rmm
rmm.reinitialize(
    managed_memory=True)
cp.cuda.set_allocator(rmm.rmm_cupy_allocator)
```
RMM lets you oversubscribe VRAM.
With that enabled I can run a pseudo batch `cuml.make_blobs(n_samples=5600000, n_features=40)` on my 3090. I don't have access to colab pro. If the error persists please let me know. I have to feeling it might have something to do with colab.

Can you try to run this in colab?
```
from anndata import AnnData
import cupy as cp
import rapids_singlecell as rsc
import scanpy as sc
import cuml

import rmm
rmm.reinitialize(
    managed_memory=True)
cp.cuda.set_allocator(rmm.rmm_cupy_allocator)

# create the dataset
blobs = cuml.make_blobs(n_samples=5600000, n_features=40)

adata = AnnData(blobs[0].get())
adata.obs["blob"] = blobs[1].get().astype(str)
del blobs

# Create the pseudo batch effect
adata.X[:2800000] = adata.X[:2800000]*2
adata.obs["batch"] = "2"
adata.obs[ "batch"][:2800000] = "1"

#pca
rsc.tl.pca(adata)
sc.pl.pca(adata,color= ["batch","blob"])

#harmony
rsc.tl.harmony_integrate(adata,key = "batch")
sc.pl.embedding(adata,color= ["batch","blob"],basis="X_pca_harmony")
```
The whole script should take around 5-6 min the most of which is harmony

**Comment by johnhickey22** [comment]
This code was super helpful. It took about 7 minutes to complete on my end and your code worked and produced the output I captured from screen shot below:

<img width="686" alt="image" src="https://github.com/Intron7/rapids_singlecell/assets/51136713/ee77c7a3-e2c7-45cb-99c0-e076889fc9a1">

However, when I run my rsc.tl.harmony_integrate() on my adata, I still get the same error even after running the code above (e.g., the rmm portion): 

---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
[<ipython-input-125-b548505bfd42>](https://localhost:8080/#) in <cell line: 1>()
----> 1 rsc.tl.harmony_integrate(adata=adata, key='file', adjusted_basis='X_pca_harmony',)
      2 # ERROR
      3 # CUTLASS error encountered at: file=/__w/cuml/cuml/python/_skbuild/linux-x86_64-3.10/cmake-build/_deps/raft-src/cpp/include/raft/distance/detail/pairwise_distance_cutlass_base.cuh line=159: call='cutlassDist_op.can_implement(arguments)', Reason=Error Invalid Problem
      4 # Obtained 64 stack frames

4 frames
[/usr/local/lib/python3.10/dist-packages/cuml/internals/api_decorators.py](https://localhost:8080/#) in wrapper(*args, **kwargs)
    186 
    187                     if process_return:
--> 188                         ret = func(*args, **kwargs)
    189                     else:
    190                         return func(*args, **kwargs)

kmeans.pyx in cuml.cluster.kmeans.KMeans.fit()

RuntimeError: CUTLASS error encountered at: file=/__w/cuml/cuml/python/_skbuild/linux-x86_64-3.10/cmake-build/_deps/raft-src/cpp/include/raft/distance/detail/pairwise_distance_cutlass_base.cuh line=159: call='cutlassDist_op.can_implement(arguments)', Reason=Error Invalid Problem
Obtained 64 stack frames
#0 in /usr/local/lib/python3.10/dist-packages/cuml/internals/../libcuml++.so(_ZN4raft9exception18collect_call_stackEv+0x81) [0x7946f805fe91]
#1 in /usr/local/lib/python3.10/dist-packages/cuml/internals/../libcuml++.so(_ZN4raft13cutlass_errorC1ERKSs+0x14b) [0x7946f85e883b]
#2 in /usr/local/lib/python3.10/dist-packages/cuml/internals/../libcuml++.so(+0x3526714) [0x7946f9643714]
#3 in /usr/local/lib/python3.10/dist-packages/cuml/internals/../libcuml++.so(_ZN4raft8distance6detail24pairwise_matrix_dispatchINS1_3ops18l2_exp_distance_opIddiEEdddNS_11identity_opEiEEvT_T4_S8_S8_PKT0_SB_SB_SB_PT2_T3_P11CUstream_stb+0x46c) [0x7946f9647a0c]
#4 in /usr/local/lib/python3.10/dist-packages/cuml/internals/../libcuml++.so(+0x32d96fc) [0x7946f93f66fc]
#5 in /usr/local/lib/python3.10/dist-packages/cuml/internals/../libcuml++.so(+0x264d7d1) [0x7946f876a7d1]
#6 in /usr/local/lib/python3.10/dist-packages/cuml/internals/../libcuml++.so(+0x268f0bc) [0x7946f87ac0bc]
#7 in /usr/local/lib/python3.10/dist-packages/cuml/internals/../libcuml++.so(+0x26985c2) [0x7946f87b55c2]
#8 in /usr/local/lib/python3.10/dist-packages/cuml/internals/../libcuml++.so(_ZN2ML6kmeans11fit_predictERKN4raft8handle_tERKNS1_7cluster6kmeans12KMeansParamsEPKdiiSB_PdPiRdRi+0xdc) [0x7946f87b666c]
#9 in /usr/local/lib/python3.10/dist-packages/cuml/cluster/kmeans.cpython-310-x86_64-linux-gnu.so(+0x3969e) [0x7947d9ea269e]
#10 in /usr/bin/python3(PyObject_Call+0xbb) [0x597cc4901b5b]
#11 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x2a37) [0x597cc48ddd87]
#12 in /usr/bin/python3(+0x16af11) [0x597cc4900f11]
#13 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x640a) [0x597cc48e175a]
#14 in /usr/bin/python3(_PyFunction_Vectorcall+0x7c) [0x597cc48f34ec]
#15 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x8c4) [0x597cc48dbc14]
#16 in /usr/bin/python3(_PyObject_FastCallDictTstate+0xc4) [0x597cc48e88b4]
#17 in /usr/bin/python3(+0x167225) [0x597cc48fd225]
#18 in /usr/bin/python3(_PyObject_MakeTpCall+0x1fc) [0x597cc48e96cc]
#19 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x67dc) [0x597cc48e1b2c]
#20 in /usr/bin/python3(_PyFunction_Vectorcall+0x7c) [0x597cc48f34ec]
#21 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x2a37) [0x597cc48ddd87]
#22 in /usr/bin/python3(_PyFunction_Vectorcall+0x7c) [0x597cc48f34ec]
#23 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x1a1b) [0x597cc48dcd6b]
#24 in /usr/bin/python3(+0x142176) [0x597cc48d8176]
#25 in /usr/bin/python3(PyEval_EvalCode+0x86) [0x597cc49cdc56]
#26 in /usr/bin/python3(+0x23d93d) [0x597cc49d393d]
#27 in /usr/bin/python3(+0x15d749) [0x597cc48f3749]
#28 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x6cd) [0x597cc48dba1d]
#29 in /usr/bin/python3(+0x17a640) [0x597cc4910640]
#30 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x287f) [0x597cc48ddbcf]
#31 in /usr/bin/python3(+0x17a640) [0x597cc4910640]
#32 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x287f) [0x597cc48ddbcf]
#33 in /usr/bin/python3(+0x17a640) [0x597cc4910640]
#34 in /usr/bin/python3(+0x25985f) [0x597cc49ef85f]
#35 in /usr/bin/python3(+0x1689fa) [0x597cc48fe9fa]
#36 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x8c4) [0x597cc48dbc14]
#37 in /usr/bin/python3(_PyFunction_Vectorcall+0x7c) [0x597cc48f34ec]
#38 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x6cd) [0x597cc48dba1d]
#39 in /usr/bin/python3(_PyFunction_Vectorcall+0x7c) [0x597cc48f34ec]
#40 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x8c4) [0x597cc48dbc14]
#41 in /usr/bin/python3(+0x16af11) [0x597cc4900f11]
#42 in /usr/bin/python3(PyObject_Call+0x122) [0x597cc4901bc2]
#43 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x2a37) [0x597cc48ddd87]
#44 in /usr/bin/python3(+0x16af11) [0x597cc4900f11]
#45 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x1a1b) [0x597cc48dcd6b]
#46 in /usr/bin/python3(+0x203b15) [0x597cc4999b15]
#47 in /usr/bin/python3(+0x15d749) [0x597cc48f3749]
#48 in /usr/bin/python3(+0x239845) [0x597cc49cf845]
#49 in /usr/bin/python3(+0x2ea682) [0x597cc4a80682]
#50 in /usr/bin/python3(+0x15062b) [0x597cc48e662b]
#51 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x6cd) [0x597cc48dba1d]
#52 in /usr/bin/python3(_PyFunction_Vectorcall+0x7c) [0x597cc48f34ec]
#53 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x8c4) [0x597cc48dbc14]
#54 in /usr/bin/python3(+0x203b15) [0x597cc4999b15]
#55 in /usr/bin/python3(+0x15d749) [0x597cc48f3749]
#56 in /usr/bin/python3(+0x239845) [0x597cc49cf845]
#57 in /usr/bin/python3(+0x2ea682) [0x597cc4a80682]
#58 in /usr/bin/python3(+0x15062b) [0x597cc48e662b]
#59 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x6cd) [0x597cc48dba1d]
#60 in /usr/bin/python3(+0x16af11) [0x597cc4900f11]
#61 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x6cd) [0x597cc48dba1d]
#62 in /usr/bin/python3(+0x203b15) [0x597cc4999b15]
#63 in /usr/bin/python3(+0x15d749) [0x597cc48f3749]

**Comment by Intron7** [comment]
I found the issue. The initialisation from from k-means++ to k-means|| and this fixed it. I'll have this fixed for the next release.

**Comment by johnhickey22** [comment]
Super this sounds great. I just tried editing locally and it worked (took about 24 min with my machine) When do you anticipate having the next release through pypi, just so I don't have to edit locally everytime?

Also, just a small minor note, the rsc.tl.leiden produces clusters as normal, but what is different from sc.tl.leiden is that sc.tl.leiden returns the clusters in the order where the first cluster (0) has the greatest number of cells assigned to it and the second number (1) has the second most number of clusters present, whereas the rsc version seems to just generate a list where the number associated with the cluster is not in order of the prevalence of the cluster. It is not a big deal (I wrote a custom script to realign these) but thought it might make transition for other sc. users easier. 

Thanks again for looking into the harmony issue for me. Really appreciate it!

**Comment by Intron7** [comment]
Fixed with version 0.7.5. Leiden clustering is broken at the moment on the rapids side. I created an issue for that.
